### PR TITLE
Feature/postgres merge events

### PR DESCRIPTION
**Pull request by akhtyamovrr**

### CODE REVIEW

File: adapters/postgres.go [hunk 1]
```diff
@@ -235,10 +277,44 @@ func (p *Postgres) patchTableSchemaInTransaction(wrappedTx *Transaction, patchSc
 			return fmt.Errorf("Error patching %s table with '%s' - %s column schema: %v", patchSchema.Name, columnName, mappedColumnType, err)
 		}
 	}
+	// PKFields of the diff set to nil when no modification required. If the value is an empty array, it means that
```
Comment (akhtyamovrr) [inline adapters/postgres.go:280]: I'm not sure if this is the best way to use here. Maybe something like Optional in java should be used here

**Review by xtreding — COMMENTED** [review]
Since primary keys don't depend on input events structure, I think, we should move primaryKeys to TableHelper and apply the following logic:

1) adapters.TableManager will have methods for deleting and adding constraints
2) in EnsureTable() we should:
   a) get current table (with primaryKeys and hold them into schema object).
   b) create table with constraints if doesn't exist
   b) if exist - manually check if the current constraint isn't equal to the current configuration call adapters.TableManager  del/add constraint methods
   c) after patching table with constraint we should increment table version as well
   

3) schema Diff() and schema Exists() shouldn't take into account primaryKeys
4) primaryKeys should be a map (comparing two maps is easier than two arrays)

The main idea is - to have patching table with columns and patching table with constraint separately in tableHelper.

File: adapters/postgres.go [hunk 2]
```diff
@@ -268,20 +344,41 @@ func (p *Postgres) InsertInTransaction(wrappedTx *Transaction, schema *schema.Ta
...
 
+func (p *Postgres) insertQuery(pkFields []string, tableName string, header string, placeholders string) string {
+	if pkFields == nil || len(pkFields) == 0 {
+		return fmt.Sprintf(insertTemplate, p.config.Schema, tableName, header, placeholders)
+	} else {
+		return fmt.Sprintf(mergeTemplate, p.config.Schema, tableName, header, placeholders, PkFieldsToConstraint(p.config.Schema, tableName), updateSection(header))
+	}
+}
+
+func PkFieldsToConstraint(schemaName string, tableName string) string {
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply adapters/postgres.go:369]: fixed

File: adapters/postgres.go [hunk 3]
```diff
@@ -268,20 +344,41 @@ func (p *Postgres) InsertInTransaction(wrappedTx *Transaction, schema *schema.Ta
...
 	if err != nil {
-		return fmt.Errorf("Error inserting in %s table with statement: %s values: %v: %v", schema.Name, header, values, err)
+		return fmt.Errorf("Error inserting in %s table with statement: %s values: %v: %v", table.Name, header, values, err)
 	}
 
 	return nil
 }
 
+func (p *Postgres) insertQuery(pkFields []string, tableName string, header string, placeholders string) string {
+	if pkFields == nil || len(pkFields) == 0 {
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply adapters/postgres.go:362]: fixed

File: adapters/postgres.go [hunk 4]
```diff
@@ -28,10 +28,24 @@ const (
...
+								pg_class.oid = $2::regclass AND
+								indrelid = pg_class.oid AND
+								nspname = $1 AND
+								pg_class.relnamespace = pg_namespace.oid AND
+								pg_attribute.attrelid = pg_class.oid AND
+								pg_attribute.attnum = any(pg_index.indkey)
+					  	AND indisprimary`
 	createDbSchemaIfNotExistsTemplate = `CREATE SCHEMA IF NOT EXISTS "%s"`
 	addColumnTemplate                 = `ALTER TABLE "%s"."%s" ADD COLUMN %s %s`
+	dropPrimaryKey                    = "ALTER TABLE %s.%s DROP CONSTRAINT IF EXISTS %s"
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply adapters/postgres.go:44]: fixed

File: adapters/postgres.go [hunk 5]
```diff
@@ -28,10 +28,24 @@ const (
   							AND  pg_namespace.nspname = $1
   							AND pg_class.relname = $2
   							AND pg_attribute.attnum > 0`
+	primaryKeyFields = `SELECT
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply adapters/postgres.go:31]: fixed

**Comment by akhtyamovrr** [comment]
@xtreding , I've implemented the approach when schema and constraints are patched separately. The only thing is that I have called primary keys patching code at one place, not at two, as you suggested in your comment.
- If a new table is created, it is created without primary keys, and then, when we compare the existing schema with the requested one, the primary key is created. 
- If a table exists, patch the existing one

From my point of view, such approach is simple and allows avoiding multiple table creation queries or building it using concatenation which simplifies logic. Also, I don't see the problem creating a primary key at a separate transaction.

For those destinations that don't have primary keys implementation yet (BigQuery, ClickHouse, Snowflake), I just write Warn log. Returning error seems too much for such a scenario.

File: schema/table.go [hunk 6]
```diff
@@ -52,6 +54,7 @@ func (t *Table) Exists() bool {
 // Return schema to add to current schema (for being equal) or empty if
 // 1) another one is empty
 // 2) all fields from another schema exist in current schema
+// 3) primary key is the same at both schemas
```
Comment (xtreding) [inline schema/table.go:57]: can't find №3 in this func

**Review by xtreding — APPROVED** [review]

File: schema/table.go (hunk 6 above)
Comment (akhtyamovrr) [reply schema/table.go:57]: You are damn right

**Pull request merged**